# Worksheet 08

Name:  Camille Christie
UID: U89708684

### Topics

- Soft Clustering
- Clustering Aggregation

### Probability Review

Read through [the following](https://medium.com/@gallettilance/overview-of-probability-3272b72c82c8)

### Soft Clustering

We generate 10 data points that come from a normal distribution with mean 5 and variance 1.

In [2]:
import random
import numpy as np
from sklearn.cluster import KMeans

mean = 5
stdev = 1

c1 = np.random.normal(mean, stdev, 10).tolist()
print(c1)

[5.856550822197005, 5.053058546056018, 4.6250737416926455, 3.850200078621048, 4.7886201319423, 3.8236845767174303, 3.743569446508812, 4.3352127348322735, 4.157596723934971, 4.172545275620438]


a) Generate 10 more data points, this time coming from a normal distribution with mean 8 and variance 1.

In [4]:
mean = 8
stdev = 1

c2 = np.random.normal(mean, stdev, 10).tolist()
print(c2)

[8.471232011694868, 8.592690535913478, 10.259127492439832, 8.792418161986022, 7.788835390722096, 10.34118432626278, 9.193499482527312, 8.651610123164799, 6.601342762722744, 6.322414449382664]


b) Flip a fair coin 10 times. If the coin lands on H, then pick the last data point of `c1` and remove it from `c1`, if T then pick the last data point from `c2` and remove it from `c2`. Add these 10 points to a list called `data`.

In [5]:
data = []
for i in range(10):
    # flip coin
    coin_output = random.choice([0, 1])
    if coin_output == 0:
        p1 = c1.pop()
        data.append(p1)
    else:
        p2 = c2.pop()
        data.append(p2)
print(data)

[4.172545275620438, 4.157596723934971, 4.3352127348322735, 3.743569446508812, 6.322414449382664, 6.601342762722744, 3.8236845767174303, 4.7886201319423, 3.850200078621048, 8.651610123164799]


c) This `data` is a Gaussian Mixture Distribution with 2 mixture components. Over the next few questions we will walk through the GMM algorithm to see if we can uncover the parameters we used to generate this data. First, please list all these parameters of the GMM that created `data` and the values we know they have.

mean, variance

d) Let's assume there are two mixture components (note: we could plot the data and make the observation that there are two clusters). The EM algorithm asks us to start with a random `mean_j`, `variance_j`, `P(C_j)` for each component j. One method we could use to find sensible values for these is to apply K means with k=2 here.

1. the centroids would be the estimates of the `mean_j`
2. the intra-cluster variance could be the estimate of `variance_j`
3. the proportion of points in each cluster could be the estimate of `P(C_j)`

Go through this process and list the parameter estimates it gives. Are they close or far from the true values?

In [7]:
kmeans = KMeans(2, init='k-means++').fit(X=np.array(data).reshape(-1, 1))

c1 = [x[0] for x in filter(lambda x: x[1] == 0, zip(data, kmeans.labels_))]
print(c1)

c2 = [x[0] for x in filter(lambda x: x[1] == 1, zip(data, kmeans.labels_))]
print(c2)

prob_c = [len(c1) / (len(c1) + len(c2)), len(c2) / (len(c1) + len(c2))]
mean = [sum(c1) / len(c1), sum(c2) / len(c2)]

var = [sum(map(lambda x: (x - mean[0]) ** 2, c1)) / len(c1), sum(map(lambda x: (x - mean[1]) ** 2, c2)) / len(c2)]

print("P(C_1) = " + str(prob_c[0]) + ", P(C_2) = " + str(prob_c[1]))
print("mean_1 = " + str(mean[0]) + ", mean_2 = " + str(mean[1]))
print("var_1 = " + str(var[0]) + ", var_2 = " + str(var[1]))

[4.172545275620438, 4.157596723934971, 4.3352127348322735, 3.743569446508812, 3.8236845767174303, 4.7886201319423, 3.850200078621048]
[6.322414449382664, 6.601342762722744, 8.651610123164799]
P(C_1) = 0.7, P(C_2) = 0.3
mean_1 = 4.124489852596753, mean_2 = 7.191789111756736
var_1 = 0.11424251421881572, var_2 = 1.0785055266713541


/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:1416: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  super()._check_params_vs_input(X, default_n_init=10)


The first mean is 4.12, the second is 7.19. The first variance is 0.11, the second is 1.07. They all seem to be close to the true value.

e) For each data point, compute `P(C_j | X_i)`. Comment on which cluster you think each point belongs to based on the estimated probabilities. How does that compare to the truth?

In [13]:
from scipy.stats import norm

prob_c0_x = [] # P(C_0 | X_i)
prob_c1_x = [] # P(C_1 | X_i)
prob_x = [] # P(X_i)

k = 2

for p in data:
    print("point = ", p)
    pdf_i = []

    for j in range(k):
        # P(X_i | C_j)
        pdf_i.append(norm.pdf(p, mean[j], var[j]))
        print("probability of observing that point if it came from cluster " + str(j) + " = ", pdf_i[j])
        # P(C_j) already computed
        prob_c[j]

    # P(X_i) = P(C_0)P(X_i | C_0) + P(C_1)P(X_i | C_1)
    prob_x = prob_c[0] * pdf_i[0] + prob_c[1] * pdf_i[1]

    # P(C_j | X_i) = P(X_i | C_j)P(C_j) / P(X_i)
    prob_c0_x.append( ... )
    prob_c1_x.append( ... )

probs = zip(data, prob_c0_x, prob_c1_x)
for p in probs:
    print(p[0])
    print("Probability of coming from C_1 = " + str(p[1]))
    print("Probability of coming from C_2 = " + str(p[2]))
    print()


point =  4.172545275620438
probability of observing that point if it came from cluster 0 =  3.196391670842603
probability of observing that point if it came from cluster 1 =  0.007350176845543776
point =  4.157596723934971
probability of observing that point if it came from cluster 0 =  3.34846745407861
probability of observing that point if it came from cluster 1 =  0.0070697594603592925
point =  4.3352127348322735
probability of observing that point if it came from cluster 0 =  0.6372212864331996
probability of observing that point if it came from cluster 1 =  0.011084912143475557
point =  3.743569446508812
probability of observing that point if it came from cluster 0 =  0.013455983002200984
probability of observing that point if it came from cluster 1 =  0.0022302906260105127
point =  6.322414449382664
probability of observing that point if it came from cluster 0 =  1.4704574168412395e-80
probability of observing that point if it came from cluster 1 =  0.2672937629204784
point =  6.

f) Having computed `P(C_j | X_i)`, update the estimates of `mean_j`, `var_j`, and `P(C_j)`. How different are these values from the original ones you got from K means? briefly comment.

In [ ]:
prob_c = [sum(prob_c0_x)/ len(prob_c0_x), ... ]
mean = [sum([x[0] * x[1] for x in zip(prob_c0_x, data)]) / sum(prob_c0_x), ... ]
var = [ ... , ... ]

print("P(C_1) = " + str(prob_c[0]) + ",  P(C_2) = " + str(prob_c[1]))
print("mean_1 = " + str(mean[0]) + ",  mean_2 = " + str(mean[1]))
print("var_1 = " + str(var[0]) + ",  var_2 = " + str(var[1]))

g) Update `P(C_j | X_i)`. Comment on any differences or lack thereof you observe.

h) Use `P(C_j | X_i)` to create a hard assignment - label each point as belonging to a specific cluster (0 or 1)

### Clustering Aggregation

| Point | C | P |
|-------|---|---|
| A     | 0 | a |
| B     | 0 | b |
| C     | 2 | b |
| D     | 1 | c |
| E     | 1 | d |

a) Fill in the following table where for each pair of points determine whether C and P agree or disagree on how to cluster that pair.

| Pair | Disagreement |
|------|--------------|
| A  B |      ?       |
| A  C |      ?       |
| A  D |      ?       |
| A  E |      ?       |
| B  C |      ?       |
| B  D |      ?       |
| B  E |      ?       |
| C  D |      ?       |
| C  E |      ?       |
| D  E |      ?       |


As datasets become very large, this process can become computationally challenging.

b) Given N points, what is the formula for the number of unique pairs of points one can create?

Assume that clustering C clusters all points in the same cluster and clustering P clusters points as such:

| Point | P |
|-------|---|
| A     | 0 |
| B     | 0 |
| C     | 0 |
| D     | 1 |
| E     | 1 |
| F     | 2 |
| G     | 2 |
| H     | 2 |
| I     | 2 |

c) What is the maximum number of disagreements there could be for a dataset of this size? (use the formula from b)?

d) If we look at cluster 0. There are (3 x 2) / 2 = 3 pairs that agree with C (since all points in C are in the same cluster). For each cluster, determine how many agreements there are. How many total agreements are there? How many disagreements does that mean there are between C and P?

e) Assuming that filtering the dataset by cluster number is a computationally easy operation, describe an algorithm inspired by the above process that can efficiently compute disagreement distances on large datasets.